### ISSUE TITLE
Empty Branch and Obsolete statements appears although they are in baseline

### ISSUE THREAD

**Issue by fidley** [larshp/abapOpenChecks#647]
Hello Lars, 
I know this is mainly not the focus of this project, but I faced some strange situation, which a bit destroys easy entry to ATC in my company. I have setup a baseline but after some change in the code somehow only empty branch (18) and obsolete statements(08) has become again errors to correct. So baseline is not considered. 

All SAP checks remained as they were in the baseline. I think that something may be missing while passing to the INFORM method, but I'm not yet sure.

Have you seen such behavior before?

**Comment by larshp** [comment]
I've not used the baseline feature, as I choose to fix the problems instead.

Anyhow, can you explain a bit more? I'm not sure I understand, empty branch and obsolete has become errors again, is this as expected or unexpected, and does this mean that these 2 works with baseline or not?

**Comment by fidley** [comment]
Hi Lars, 

so here is the story in detail:
I was introducing ATC in my company, in which we are using SAP since R/2 I think. So there are some programs > 20 years old, done in bad way, but we have to maintain some changes still there. Also some of my colleagues doesn't use OO and they stayed behind a long time ago. Also myself have done really bad code as I was learning to develop in ABAP from the code that was already existing in company or SCN. So while introducing ATC I needed to use baseline, as in other case I would fully stopped the developments of my colleagues, which was of course not welcomed by their managers. 

But what was strange is that, although I have setup the baseline, then still old findings were appearing as new one... I thought at the beginning that this is SAP error, so I've searched for some OSS notes and I found some. I've installed them and for the SAP checks it helped (the reason was that sometimes when ATC found more than one same error, then not all of the checks were saved as a finding). But now more on more I got information from the devs that again they're receiving the errors that should be in the baseline, although they've changed only one or two lines. The SAP checks remained found by the baseline. 

So for me this is unexpected behavior and it makes the acceptance of ATC really low...

I've checked a bit the SAP checks and it seems that they pass more information to the INFORM method, one crucial could be CHECKSUM and/or COLUMN  and/or INCLSPEC

![image](https://user-images.githubusercontent.com/7912195/51340067-4a8ddc80-1a8e-11e9-956e-e3eb76564f6d.png)

Do you have idea about this fields?

**Comment by larshp** [comment]
havent tried the fields, but seen it used in some of the standard checks

will need to do some testing, sometime :)

https://help.sap.com/viewer/ba879a6e2ea04d9bb94c7ccd7cdac446/7.51.4/en-US/21f52fff7cef458b88e49c6087895162.html

**Comment by fidley** [comment]
This seems to be most important
![image](https://user-images.githubusercontent.com/7912195/51341757-5d0a1500-1a92-11e9-8eb6-2667875015cf.png)
But I think the checksum, column  or inclspec made it that the scope is kept (+/- some lines above/bellow). If I would bet it would be checksum :)

**Comment by fidley** [comment]
I'm trying now to add this all fields to the checks, let's see if it will change something

**Comment by fidley** [comment]
It may be the problem of all checks not only this two, but this two are set as an error in my case and they are very common. 

I've made the changes on our remote ATC system. I'll let you know if it helped. If yes I will tell you how I've done it for this two checks, so the others could be also adapted.

**Comment by oliver-huetkoeper** [comment]
Hi,

we are facing the same issue. Therefore I have analyzed it. The baseline depends on the identity/checksum of a finding. As long as it is the same, the base line recognizes it.

Unfortunately currently none of the open checks handles the checksum generation. In this case a fallback is implemented in the ATC logic. This checksum is based on the line number of the finding. And as soon as you insert/remove code above this line, the base line cannot match it.

Luckily, we can trigger the checksum generation easily within a check class. We need to do two things:
1. Set the uses_checksum attribute in the constructor method to abap_true
2. Fill parameter p_position in the "inform" method call with the index of the affected statement (index in table it_statements) 

I have implemented this for some checks (including empty branch and obsolete statements) and create a pull request.

Best regards,
Oliver
